### ISSUE TITLE
DUCT example incorrectly combines all players moves in node selection and final move selection

### ISSUE THREAD

**Issue by EklipZgit** [Ludeme/LudiiExampleAI#8]
Reference this commit for the fix in my python port:
https://github.com/EklipZgit/generals-bot/commit/9db32b58ec8050d589b121d2a9510cf252967bda

I'm not 100% sure this is a problem in Ludii as it is difficult to tell how that BoardMoves class is being used in the SM implementation and I'm not actually running Ludii. However the code appears to pretty clearly be combining all player possible moves into a set of board moves that each player then iterates through, resulting in exploring the tree based on optimistic opponent behavior, when in fact it needs to explore based on its own moves alone and let the opponent choose their own move.

Ludii ExampleDUCT KIND OF does this by separating the utilities, but doesn't separate the moves, resulting in very optimistic move selection with regard to opponent response. After I made the above change in my port, DUCT immediately started spitting out correct opponent behavior.

--------
EDIT, might be I misinterpreted what
```
AIUtils.extractMovesForMover(allLegalMoves, p)
```
does, and how the Ludii implementation interacts with that. But the Move class clearly has all players actions combined in one entry, which is how I ported the implementation and had the problem. You'll want to make sure AIUtils.extractMovesForMover(allLegalMoves, p) does not include the opponents contribution and that finalMoveSelection finds the best move for each player at each node and combines them into the combined set of moves.

**Comment by DennisSoemers** [comment]
I'm a bit confused. Did you at any point ever actually run any Ludii code and identify an issue there? Or did you just attempt to follow a similar style of coding / API in a different framework, and only actually get a bug there?

**Comment by EklipZgit** [comment]
I have not reproduced the issue in Ludii as I don't know how to get a SM game set up there. I was just porting the MCTS DUCT implementation to python and ran into these issues.

If you tell me how to run a SM game in Ludii, I'll attempt to repro the issue and PR the fix. The fact that it appears to be using the same Move class for a combination of all players moves, as well as the individual player contributions to a given 'simultaneous move' makes me confident this is an issue in Ludii, but I was too lazy to spend hours trying to figure out how to get a Ludii repro set up since the Game / Moves stuff is so abstract I wasn't able to follow it backwards to the exact concrete implementations a SM game in Ludii would use. It certainly could be I'm misunderstanding something about how Ludii stores the combination moves vs player moves since the implementation and application of those are both abstract and I was never able to find the source for their implementations for SM move games.

If you give me a few links to a few blocks of instructions that would let me get a Ludii SM game running with a breakpoint in the DUCT in under an hour, I'll happily poke in there to confirm whether I'm correct or not.

The real question just boils down to: Do the `Move`s returned by AIUtils.extractMovesForMover(allLegalMoves, p) include the opponent contributions?

EG lets say game is two player rock paper scissors, allLegalMoves is
(p1, p2)
rock, rock
rock, paper
rock, scissors
paper, rock
paper, paper
paper, scissors
scissors, rock
scissors, paper
scissors, scissors

if AIUtils.extractMovesForMover(allLegalMoves, 1) returns 9 moves instead of 3 moves (rock, paper, scissors), then I'm 99.99% sure you have the bug I am describing. If AIUtils.extractMovesForMover(allLegalMoves, 1) returns 3 moves and AIUtils.extractMovesForMover(allLegalMoves, 2) returns 3 moves, then you're fine and I was just super confused by both allLegalMoves and result of AIUtils.extractMovesForMover(allLegalMoves, 1) returning the same Move abstraction but sometimes using it to mean multiple agents combined choices or not.


Rock paper scissors might not be the best example here though because you could also in theory just say 'allLegalMoves' was just 'rock, paper, scissors' already since all players would mirror the same moves.

Lets imagine 4 players playing free-for-all Chess with simultaneous moves, with only a single king left on the board each in one of the corners (so each player has two possible moves on our example turn, but one of the players might have 3, or 4 possible moves on any given turn, or might have a rook left with lots of options, or whatever. But for this example, 1 king in each corner).
I'll annotate the player move options with p{playerNumber}-{moveOption} where move option indicates which of the two possible king moves they make from this board state.

If I'm reading Ludii right, then I expect allLegalMoves to contain all legal combinations of player contributions to a given simultaneous move, correct?
List<Move>
(p1, p2, p3, p4)
p1-0, p2-0, p3-0, p4-0
p1-0, p2-0, p3-0, p4-1
p1-0, p2-0, p3-1, p4-0
p1-0, p2-0, p3-1, p4-1
p1-0, p2-1, p3-0, p4-0
p1-0, p2-1, p3-0, p4-1
p1-0, p2-1, p3-1, p4-0
p1-0, p2-1, p3-1, p4-1
p1-1, p2-0, p3-0, p4-0
p1-1, p2-0, p3-0, p4-1
p1-1, p2-0, p3-1, p4-0
p1-1, p2-0, p3-1, p4-1
p1-1, p2-1, p3-0, p4-0
p1-1, p2-1, p3-0, p4-1
p1-1, p2-1, p3-1, p4-0
p1-1, p2-1, p3-1, p4-1

Then after allLegalMoves is populated with those 16 options, what exactly does 
AIUtils.extractMovesForMover(allLegalMoves, 1) output? If the answer isn't 
List<Move>
(p1)
p1-0
p1-1

then you have the bug. If it IS that, then I completely misunderstood how DUCT was using the Moves class differently in those two contexts (or I misunderstood what allLegalMoves contains and it instead assumes each player has all the same options as each other player, which would prevent this algo from working for games like chess with asymmetric move lists available per player, so I've also been assuming that that's not the case).

Given that building the legal move list is done for the board state, not the players, I am reasonably confident that my interpretation of the Ludii code is correct and the issue exists. But could be I'm totally wrong and it's doing some fancy magic inside those two methods I mention to do something other than what it appears to be doing.

**Comment by DennisSoemers** [comment]
The `allLegalMoves` returned by `game.moves(context).moves()` does not return composite moves. It just returns a flat list of all individual moves, each of which is legal for exactly one of the players. So, in the initial game state of Rock-Paper-Scissors, it returns a list of six (neither 3 nor 9 as you suggest, but 6) moves (three for each player):

> allLegalMoves = 
> [[Move:typeFrom=Cell,from=2,typeTo=Cell,to=0,decision=true], 
> [Move:typeFrom=Cell,from=3,typeTo=Cell,to=0,decision=true], 
> [Move:typeFrom=Cell,from=4,typeTo=Cell,to=0,decision=true], 
> [Move:typeFrom=Cell,from=5,typeTo=Cell,to=1,decision=true], 
> [Move:typeFrom=Cell,from=6,typeTo=Cell,to=1,decision=true], 
> [Move:typeFrom=Cell,from=7,typeTo=Cell,to=1,decision=true]]

Then, when we call `AIUtils.extractMovesForMover(allLegalMoves, p)`, it returns just 3 individual moves for each player:

> AIUtils.extractMovesForMover(allLegalMoves, 1) = 
> [[Move:typeFrom=Cell,from=2,typeTo=Cell,to=0,decision=true], 
> [Move:typeFrom=Cell,from=3,typeTo=Cell,to=0,decision=true], 
> [Move:typeFrom=Cell,from=4,typeTo=Cell,to=0,decision=true]]

> AIUtils.extractMovesForMover(allLegalMoves, 2) = 
> [[Move:typeFrom=Cell,from=5,typeTo=Cell,to=1,decision=true], 
> [Move:typeFrom=Cell,from=6,typeTo=Cell,to=1,decision=true], 
> [Move:typeFrom=Cell,from=7,typeTo=Cell,to=1,decision=true]]

So, in conclusion, everything seems to be fine here and there is no bug (apart from an unrelated crash I just fixed).

**Comment by EklipZgit** [comment]
ok, sweet. Sorry for the false alarm, then!
